### PR TITLE
remove goto_programt::instructiont::function member [blocks: #3113]

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/goto_program.h [hunk 1]
```diff
@@ -424,42 +420,6 @@ class goto_programt
...
-      ++l;
-
-    return l->function;
-  }
-
-  /// Get the id of the function that contains the given goto program.
-  ///
-  /// \param p: the goto program
-  /// \return id of the function that contains the goto program
-  static const irep_idt get_function_id(
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h:566]: This function is actually used and thus any of its call-sites need to be modified.

**Review by tautschnig — DISMISSED** [review]
If we can actually do this then that's great, but it will require some more work.

**Comment by martin-cs** [comment]
@chrisr-diffblue : something gives me a feeling this may break a branch you maintain for a customer.  You might want to check.

**Comment by kroening** [comment]
@tautschnig There will be a few places where the name of the function currently looked at will have to be passed down the call-tree; but I don't expect that many.

**Comment by tautschnig** [comment]
Several commits have been factored out into PRs of their own, and some more of those are to be done - but they have dependencies and thus I'd rather await the merge of the commits factored out thus far. The factored-out PRs also include fixes such as missing Doxygen comments.

**Review by smowton — APPROVED** [review]
Double check: is there a test that goto-inline isn't misattributing code to the wrong function now its `adjust_function` parameter has gone away?

**Comment by tautschnig** [comment]
@kroening Even though this is your own PR it might be good for you to take a look as I have added commits of my own.

**Comment by tautschnig** [comment]
> Double check: is there a test that goto-inline isn't misattributing code to the wrong function now its `adjust_function` parameter has gone away?

Fair point - but what is right vs wrong here? The only place we could now judge this is on the source location, so what is it that we would expect? Do we want the source location to be updated, or should it stay the same?

**Comment by tautschnig** [comment]
@danpoe Would you have comments on what @smowton just raised?

**Comment by chrisr-diffblue** [comment]
This PR will break a private branch we have previously maintained for a customer, but that branch is not currently being actively maintained - so if that branch is re-awakened at any point we'll have to deal with it then. 

Regarding the question of whether source location should be updated or not, I think most compilers I've come across still report the 'original' source location for inlined code, so that seems like a sensible default behaviour. But could we not just change the 'adjust_function' parameter to a 'adjust_source_location' parameter and leave the choice of whether to modify the source_location to the caller/user ?

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f42a7bde4e04958b1e65e3addf629761b7ef7144).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99110556
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by tautschnig** [comment]
> But could we not just change the 'adjust_function' parameter to a 'adjust_source_location' parameter and leave the choice of whether to modify the source_location to the caller/user ?

Done, completely replaced the previous commit that removed `adjust_function`.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 0bd92f8cbdcd3c948637a5a7943365f3215ee941).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99193849
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by danpoe** [comment]
The inliner will continue to work as is with the function field having been removed. AFAIK previously when cbmc was still doing inlining the default behavior was for the function field of an inlined instruction to retain its original value. I think this was necessary as loop ids are pairs (function, loop number) and thus if the user specified an unwind bound for a loop that later got inlined it would not have recognised the loop otherwise.

We have some similar functionality in a private branch where we compute loop bounds on the original program, then inline everything, and then use the previously computed bounds for unwinding. We'll probably have to add some feature to translate unwind bounds given a log of inlined instructions output by the inliner.

**Review by danpoe — APPROVED** [review]

**Review by tautschnig — APPROVED** [review]
Approving, but I note that this PR includes commits of my own.

**Review by chrisr-diffblue — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 0e9e01e8698f6045c36c1e37190ea9e0df2897e8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99282323
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by tautschnig** [comment]
It seems this finally passes on CBMC and has all the approvals, but as expected it fails with TG. @allredj @peterschrammel @thk123 Could you please figure out whether that work can be scheduled? I have no idea how much effort that is...

**Comment by thk123** [comment]
Creating a bump for this now.

**Comment by thk123** [comment]
@kroening Could I get this rebased to create the bump.

**Comment by tautschnig** [comment]
@thk123 rebased.

**Comment by thk123** [comment]
@tautschnig And now failing CI here :disappointed:

**Comment by tautschnig** [comment]
Err, sure ... @thk123 should be fixed now.

**Comment by thk123** [comment]
TG bump created - will let you know when it passes.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5c5f9fe3385a66a6bd9f832604fcb3084cb8a304).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99459448
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by thk123** [comment]
TG bump is passing on this :+1:

**Comment by tautschnig** [comment]
Thanks @thk123, I will rebase and then merge.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 8ddd9ba7f72588bf45440872eb821c4e785866ef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99659348
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**